mergin/client_pull.py

```diff
@@ -551,3 +569,164 @@ def pull_project_finalize(job):
+        if file["path"] == file_path:
+            file['version'] = version
+            items = _download_items(file, temp_dir)
+            task = UpdateTask(file['path'], items, output_file)
+            download_list.extend(task.download_queue_items)
+            for item in task.download_queue_items:
+                total_size += item.size
+            update_tasks.append(task)
+            break
+    if not download_list:
+        warn = f"No {file_path} exists at version {version}"
+        mc.log.warning(warn)
+        raise ClientError(warn)
+
+    mc.log.info(f"will download file {file_path} in {len(download_list)} chunks, total size {total_size}")
+    job = DownloadJob(
+        project_path, total_size, version, update_tasks, download_list, temp_dir, None, mc, project_info
+    )
+    job.executor = concurrent.futures.ThreadPoolExecutor(max_workers=4)
+    job.futures = []
+    for item in download_list:
+        future = job.executor.submit(_do_download, item, mc, None, project_path, job)
+        job.futures.append(future)
+
+    return job
+
+
+def download_file_finalize(job):
+    """
+    To be called when download_project_file is finished
+    """
+    job.executor.shutdown(wait=True)
+
+    # make sure any exceptions from threads are not lost
+    for future in job.futures:
+        if future.exception() is not None:
+            raise future.exception()
+
+    job.mc.log.info("--- download finished")
+
+    for task in job.update_tasks:
+        task.apply(job.directory, job.mp)
+
+
+def download_project_file_diffs(mc, project_directory, file_path, versions):
+    """
+    Starts background download project file diffs for specified versions.
+    Returns handle to the pending download.
+    """
+    try:
+        mp = MerginProject(project_directory)
+    except InvalidProject as e:
+        mc.log.error(f"Couldn't create Mergin project for directory: {project_directory}: {repr(e)}")
+        return None
+    project_path = mp.metadata["name"]
+    mp.log.info(f"--- version: {mc.user_agent_info()}")
+    mp.log.info(f"--- start download diffs for {file_path}, versions: {[v for v in versions]}")
+
+    try:
+        server_info = mc.project_info(project_path)
+    except ClientError as err:
+        mp.log.error("Error getting project info: " + str(err))
+        mp.log.info("--- pull aborted")
+        raise
+
+    version_numbers = sorted([int(v[1:]) for v in versions])
+    temp_dir = os.path.join(tempfile.gettempdir(), "mergin_temp")
+    os.makedirs(temp_dir, exist_ok=True)
+    fetch_files = []
+
+    for version, version_data in versions.items():
+        if int(version[1:]) == version_numbers[0]:
+            continue
+        diff_file = copy.deepcopy(version_data)
+        diff_file['version'] = version
+        diff_file['diff'] = version_data['diff']
+        fetch_files.append(diff_file)
+
+    files_to_merge = []  # list of FileToMerge instances
+    for file in fetch_files:
```

Comment:
Do we need to keep this and the previous for-cycles separate? it looks to me it could be just one for-cycle...